Make accessKey and secretKey optional and add support to use IRSA Iam role

Anany-k opened the pull request:
## Context
Currently `accesskey` and `secretkey` are required config needed for integration with Object Storage

## Proposed solution

Minio supports IRSA EKS role already.

## Related issues

<!-- If suggesting a new feature or change, please discuss it in an issue first -->
<!-- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!-- If this does not solve entirely the issue, make also a checklist of what is done or not: -->

## Has this been tested?

<!-- Put an `x` in the box that applies: -->

- [ ] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [X] 🙋 no, because I need help 
I have only tested this when running on EKS configured with IRSA role. Will need guidance to make this MR mergeable.

## Screenshots / Screencasts

Anany-k committed:
Make accessKey and secretKey optional and add support to use IRSA Iam role

fflorent commented on app/server/lib/configureMinIOExternalStorage.ts:
```diff
@@ -1,6 +1,7 @@
 import {wrapWithKeyMappedStorage} from 'app/server/lib/ExternalStorage';
 import {appSettings} from 'app/server/lib/AppSettings';
 import {MinIOExternalStorage} from 'app/server/lib/MinIOExternalStorage';
+import {IamAwsProvider} from 'minio/dist/main/IamAwsProvider.js';
```
Nit: please remove the `.js` extension

```suggestion
import {IamAwsProvider} from 'minio/dist/main/IamAwsProvider';
```

fflorent commented on app/server/lib/MinIOExternalStorage.ts:
```diff
@@ -53,8 +53,10 @@ export class MinIOExternalStorage implements ExternalStorage {
-      secretKey: string,
+      accessKey?: string,
+      secretKey?: string,
+      credentialsProvider?: any,
```
Could you type this property?

Import `IamAwsProvider` like in `app/server/lib/configureMinIOExternalStorage.ts` and then:

```suggestion
      credentialsProvider?: IamAwsProvider,
```

fflorent commented on app/server/lib/configureMinIOExternalStorage.ts:
```diff
@@ -40,14 +41,15 @@ export function checkMinIOExternalStorage() {
     envVar: ['GRIST_DOCS_MINIO_SECRET_KEY'],
     censor: true,
   });
+  const credentialsProvider = new IamAwsProvider({});
```
What happens if we want to use accessKey + secretKey and not IRSA?

fflorent reviewed (commented):
Hello,

Thank you for your contribution

It sounds like there exist tests that are failing after your work.

[This documentation for developers](https://github.com/gristlabs/grist-core/blob/main/documentation/develop.md) is a guide to help you, take a look and feel free to ask questions (we think there are rooms for improvements)

```
  13) UsersManager
       multi-worker
         "before all" hook in "multi-worker":
     Error: Unable to get credentials: Error: Failed to get Credentials: TypeError [ERR_INVALID_CHAR]: Invalid character in header content ["X-aws-ec2-metadata-token"]
      at Client.checkAndRefreshCreds (node_modules/minio/dist/main/internal/client.ts:541:15)
      at processTicksAndRejections (node:internal/process/task_queues:105:5)
      at Client.makeRequestStreamAsync (node_modules/minio/dist/main/internal/client.ts:706:5)
      at Client.getBucketVersioning (node_modules/minio/dist/main/internal/client.ts:2159:21)
      at MinIOExternalStorage.hasVersioning (app/server/lib/MinIOExternalStorage.ts:164:24)
      at checkMinIOBucket (app/server/lib/configureMinIOExternalStorage.ts:74:8)
      at CoreCreate.checkBackend (app/server/lib/ICreate.ts:176:9)
      at FlexServer.addDoc (app/server/lib/FlexServer.ts:1463:7)
      at MergedServer.run (app/server/MergedServer.ts:198:9)
      at MergedServer.run (app/server/MergedServer.ts:220:11)
      at Context.<anonymous> (test/gen-server/lib/homedb/UsersManager.ts:1220:7)
     Caused by: Error: Failed to get Credentials: TypeError [ERR_INVALID_CHAR]: Invalid character in header content ["X-aws-ec2-metadata-token"]
      at IamAwsProvider.fetchCredentials (node_modules/minio/dist/main/IamAwsProvider.ts:91:13)
      at processTicksAndRejections (node:internal/process/task_queues:105:5)
      at IamAwsProvider.getCredentials (node_modules/minio/dist/main/IamAwsProvider.ts:60:27)
      at Client.checkAndRefreshCreds (node_modules/minio/dist/main/internal/client.ts:536:33)
      at Client.makeRequestStreamAsync (node_modules/minio/dist/main/internal/client.ts:706:5)
      at Client.getBucketVersioning (node_modules/minio/dist/main/internal/client.ts:2159:21)
      at MinIOExternalStorage.hasVersioning (app/server/lib/MinIOExternalStorage.ts:164:24)
      at checkMinIOBucket (app/server/lib/configureMinIOExternalStorage.ts:74:8)
      at CoreCreate.checkBackend (app/server/lib/ICreate.ts:176:9)
      at FlexServer.addDoc (app/server/lib/FlexServer.ts:1463:7)
      at MergedServer.run (app/server/MergedServer.ts:198:9)
      at MergedServer.run (app/server/MergedServer.ts:220:11)
      at Context.<anonymous> (test/gen-server/lib/homedb/UsersManager.ts:1220:7)
     Caused by: TypeError: Invalid character in header content ["X-aws-ec2-metadata-token"]
      at ClientRequest.setHeader (node:_http_outgoing:703:3)
      at new ClientRequest (node:_http_client:302:14)
      at Object.request (node:http:102:10)
      at /home/runner/work/grist-core/grist-core/node_modules/minio/dist/main/internal/request.ts:17:34
      at new Promise (<anonymous>)
      at request (node_modules/minio/dist/main/internal/request.ts:16:10)
      at IamAwsProvider.getIamRoleName (node_modules/minio/dist/main/IamAwsProvider.ts:184:30)
      at IamAwsProvider.getIamRoleNamedUrl (node_modules/minio/dist/main/IamAwsProvider.ts:166:33)
      at IamAwsProvider.fetchCredentials (node_modules/minio/dist/main/IamAwsProvider.ts:86:26)
      at processTicksAndRejections (node:internal/process/task_queues:105:5)
      at IamAwsProvider.getCredentials (node_modules/minio/dist/main/IamAwsProvider.ts:60:27)
      at Client.checkAndRefreshCreds (node_modules/minio/dist/main/internal/client.ts:536:33)
      at Client.makeRequestStreamAsync (node_modules/minio/dist/main/internal/client.ts:706:5)
      at Client.getBucketVersioning (node_modules/minio/dist/main/internal/client.ts:2159:21)
      at MinIOExternalStorage.hasVersioning (app/server/lib/MinIOExternalStorage.ts:164:24)
      at checkMinIOBucket (app/server/lib/configureMinIOExternalStorage.ts:74:8)
      at CoreCreate.checkBackend (app/server/lib/ICreate.ts:176:9)
      at FlexServer.addDoc (app/server/lib/FlexServer.ts:1463:7)
      at MergedServer.run (app/server/MergedServer.ts:198:9)
      at MergedServer.run (app/server/MergedServer.ts:220:11)
      at Context.<anonymous> (test/gen-server/lib/homedb/UsersManager.ts:1220:7)
```

Anany-k committed:
Add support for IRSA role with object store

Anany-k committed:
Fix test

Anany-k committed:
Remove .js extension when importing

Anany-k committed:
revert

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Add an env var GRIST_DOCS_MINIO_AUTH_AUTO_DETECT to when should try loading cred for IRSA

Anany-k commented on app/server/lib/MinIOExternalStorage.ts:
```diff
@@ -53,13 +56,17 @@ export class MinIOExternalStorage implements ExternalStorage {
   ) {
+    if (!options.accessKey || !options.secretKey) {
+      options.credentialsProvider = new IamAwsProvider({});
+    }
```
I added a env var `GRIST_DOCS_MINIO_AUTH_AUTO_DETECT` to try and load the credentials according to credential provider change if `GRIST_DOCS_MINIO_AUTH_AUTO_DETECT: true`

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Update how minio client is initliazed in the constructor

Anany-k commented on test/server/lib/MinIOExternalStorage.ts:
```diff
@@ -29,18 +29,33 @@ describe("MinIOExternalStorage", function () {
+    ) {
+      super(bucket, options, batchSize);
+      if (mockClient) {
+        (this as any)._s3 = mockClient;
```
Modified the `MinIOExternalStorage` a bit, so we don't need this

Anany-k commented on app/server/lib/MinIOExternalStorage.ts:
```diff
@@ -5,6 +5,7 @@ import {IncomingMessage} from 'http';
 import * as fse from 'fs-extra';
 import * as minio from 'minio';
 import * as stream from 'node:stream';
+import {IamAwsProvider} from 'minio/dist/main/IamAwsProvider.js';
```
(Not a js/ts expert)
I was not able to import from minio directly, looks like it does not export `IamAwsProvider`.
Also, when remove `.js` and just use `import {IamAwsProvider} from 'minio/dist/main/IamAwsProvider;` I get the below error.
```
node:internal/modules/cjs/loader:657
      throw e;
      ^

Error: Cannot find module '/grist/node_modules/minio/dist/main/IamAwsProvider'
    at createEsmNotFoundErr (node:internal/modules/cjs/loader:1419:15)
    at finalizeEsmResolution (node:internal/modules/cjs/loader:1408:15)
    at resolveExports (node:internal/modules/cjs/loader:650:14)
    at Function._findPath (node:internal/modules/cjs/loader:717:31)
    at Function._resolveFilename (node:internal/modules/cjs/loader:1369:27)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1025:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1030:22)
    at Function._load (node:internal/modules/cjs/loader:1192:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:237:24)
    at Module.require (node:internal/modules/cjs/loader:1463:12)
    at require (node:internal/modules/helpers:147:16)
    at Object.<anonymous> (/grist/_build/app/server/lib/MinIOExternalStorage.js:39:26)
    at Module._compile (node:internal/modules/cjs/loader:1706:14)
    at Object..js (node:internal/modules/cjs/loader:1839:10)
    at Module.load (node:internal/modules/cjs/loader:1441:32) {
  code: 'MODULE_NOT_FOUND',
  path: '/grist/node_modules/minio'
}

Node.js v22.19.0
```

Anany-k committed:
Fix eslint warning

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Rename authAutoDetect to useAwsCredentialChain for clarity as it uses IRSA, ECS task roles, EC2 instance metadata via MinIO's IamAwsProvider

Anany-k committed:
Merge branch 'gristlabs:main' into main

Anany-k committed:
Upgrade minio to 8.0.7 and import IamAwsProvider from package export

minio 8.0.7 re-exports IamAwsProvider from its public entrypoint, so we
no longer need to reach into 'minio/dist/main/IamAwsProvider.js' (which
also failed to resolve at runtime without the '.js' extension). Import it
from 'minio' directly, addressing review feedback on PR #1716.

Anany-k committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	app/server/lib/MinIOExternalStorage.ts
#	app/server/lib/configureMinIOExternalStorage.ts
